### PR TITLE
feat(tedapi): merge lifetime energy accumulators into /api/meters/aggregates

### LINKED ISSUE #221: "energy_exported" and "energy_imported" return zero always
I have a Raspberry Pi set up to access my Powerwall 3's wifi and my home wifi (intending to pass data to Home Assistant)
But the "energy_exported" and "energy_imported" remain 0 both in calls to `pw.poll("/api/meters/aggregates")` and to `pw.solar(verbose=True)` etc.

Any ideas?

ETA: I'm sure this isn't an issue with pypowerwall, but hoping you clever people can help me!

```
pi@raspberrypi:~ $ python3 powerwall-poll.py
Polling Tesla Powerwall 3
Site Name: 64 HPA - Firmware: 25.26.1 2c4bb00e - DIN: 1707000-30-K--TG1243490010CC

        00:00:00,       0.532 s,        ok,    pw.power() Combined power metrics (W):
{"site": 138, "solar": 225, "battery": 1, "load": 363.75}

        00:00:00,       0.000 s,        ok,    pw.level() Battery power level:
23.07692307692308

        00:00:00,       0.239 s,        ok,    pw.poll("/api/meters/aggregates"):

{
  "site": {
    "last_communication_time": "2025-09-28T12:07:52+01:00",
    "instant_power": 138,
    "instant_reactive_power": 0,
    "instant_apparent_power": 0,
    "frequency": 0,
    "energy_exported": 0,
    "energy_imported": 0,
    "instant_average_voltage": 242,
    "instant_average_current": 0.5702479338842975,
    "i_a_current": 3.487,
    "i_b_current": 0,
    "i_c_current": 0,
    "last_phase_voltage_communication_time": "0001-01-01T00:00:00Z",
    "last_phase_power_communication_time": "0001-01-01T00:00:00Z",
    "last_phase_energy_communication_time": "0001-01-01T00:00:00Z",
    "timeout": 1500000000,
    "num_meters_aggregated": 1,
    "instant_total_current": 0.5702479338842975,
    "disclaimer": "site: voltage/current from Meter X"
  },
  "battery": {
    "last_communication_time": "2025-09-28T12:07:52+01:00",
    "instant_power": 1,
    "instant_reactive_power": 0,
    "instant_apparent_power": 0,
    "frequency": 0,
    "energy_exported": 0,
    "energy_imported": 0,
    "instant_average_voltage": 240.8,
    "instant_average_current": 0.00415282392026578,
    "i_a_current": 0,
    "i_b_current": 0,
    "i_c_current": 0,
    "last_phase_voltage_communication_time": "0001-01-01T00:00:00Z",
    "last_phase_power_communication_time": "0001-01-01T00:00:00Z",
    "last_phase_energy_communication_time": "0001-01-01T00:00:00Z",
    "timeout": 1500000000,
    "num_meters_aggregated": 1,
    "instant_total_current": 0.00415282392026578,
    "disclaimer": "battery: voltage from PINV, calculated current from power"
  },
  "load": {
    "last_communication_time": "2025-09-28T12:07:52+01:00",
    "instant_power": 363.75,
    "instant_reactive_power": 0,
    "instant_apparent_power": 0,
    "frequency": 0,
    "energy_exported": 0,
    "energy_imported": 0,
    "instant_average_voltage": 241,
    "instant_average_current": 1.5093360995850622,
    "i_a_current": 0,
    "i_b_current": 0,
    "i_c_current": 0,
    "last_phase_voltage_communication_time": "0001-01-01T00:00:00Z",
    "last_phase_power_communication_time": "0001-01-01T00:00:00Z",
    "last_phase_energy_communication_time": "0001-01-01T00:00:00Z",
    "timeout": 1500000000,
    "instant_total_current": 1.5093360995850622,
    "num_meters_aggregated": 1,
    "disclaimer": "load: voltage from ISLAND_AcMeasurements, current calculated from power"
  },
  "solar": {
    "last_communication_time": "2025-09-28T12:07:52+01:00",
    "instant_power": 225,
    "instant_reactive_power": 0,
    "instant_apparent_power": 0,
    "frequency": 0,
    "energy_exported": 0,
    "energy_imported": 0,
    "instant_average_voltage": 240.60000000000002,
    "instant_average_current": 0.9351620947630922,
    "i_a_current": 3.5045,
    "i_b_current": 0,
    "i_c_current": 0,
    "last_phase_voltage_communication_time": "0001-01-01T00:00:00Z",
    "last_phase_power_communication_time": "0001-01-01T00:00:00Z",
    "last_phase_energy_communication_time": "0001-01-01T00:00:00Z",
    "timeout": 1000000000,
    "num_meters_aggregated": 1,
    "instant_total_current": 0.9351620947630922,
    "disclaimer": "solar: voltage from PVAC, calculated current from power"
  }
}

```

### PR DESCRIPTION
**Pull request by jasonacox-sam**
## Summary

PW3 firmware still serves the classic `/api/meters/aggregates` endpoint behind the **customer login** (`POST /api/login/Basic` → Bearer token) — and unlike the TEDAPI payloads, it carries the lifetime `energy_imported` / `energy_exported` accumulators for site, battery, load and solar. This was surfaced by @cavedon in #221 and confirmed here on real hardware.

This PR makes the TEDAPI adapter fetch that native payload and overlay the two energy fields onto the synthesized Powerwall-style aggregates response:

* New customer-login session in the `TEDAPI` class (`_customer_login` / `get_native_api` / `get_native_meters_aggregates`) — password is auto-derived (last 5 chars of the gateway password, same derivation the library already uses), works in both **v1r wired-LAN** mode and **WiFi full-tedapi** mode, with WiFi-host fallback.
* `get_api_meters_aggregates()` now merges `energy_imported` / `energy_exported` per section when the native endpoint is reachable, and annotates the section `disclaimer` (`…; energy from gateway local API`) so the provenance is visible.
* Gateways without the endpoint degrade gracefully — values stay `0` exactly as before, with a 5-minute retry backoff so unsupported firmware isn't hammered. Successful fetches are cached on the normal poll cadence.

## Hardware verification (Powerwall 3, firmware 26.x)

```
pw.poll("/api/meters/aggregates")  # v1r mode
site     instant_power=-51    energy_exported=1469391    energy_imported=4902666.25
battery  instant_power=1455   energy_exported=4712126    energy_imported=4803385
load     instant_power=1425   energy_imported=11679089
solar    instant_power=0      energy_exported=8337073
```

The lifetime values cross-check: load (11.68 MWh) ≈ solar (8.34) + grid import (4.90) − grid export (1.47) − battery net (−0.09). Values tick upward between polls — live accumulators, not stubs. Same result in WiFi full-tedapi mode. Full library test suite passes (277 passed, 10 skipped).

Fixes #221

## Notes

* Battery energy counters appear to be lifetime totals (Wh) as on PW2 — same semantics, so dashboards graphing daily energy will need to delta them as they already do for PW2.
* Older PW3 firmware that truly lacks the endpoint simply keeps prior behavior.

### CODE REVIEW

**Comment by jasonacox** [comment]
Excellent! Good addition @jasonacox-sam  - Please address github review,  increment the version and add entry to RELEASE.md in prep for merge.

File: pypowerwall/tedapi/pypowerwall_tedapi.py [hunk 1]
```diff
@@ -408,6 +408,27 @@ def get_api_meters_aggregates(self, **kwargs) -> Optional[Union[dict, list, str,
         for section in (data['site'], data['load'], data['solar'], data['battery']):
             section['last_communication_time'] = timestamp
 
+        # Merge lifetime energy accumulators from the gateway's local API when
+        # available. PW3 firmware still serves /api/meters/aggregates behind the
+        # customer login and it carries energy_imported/energy_exported values
+        # that the TEDAPI payloads do not (see issue #221).
+        native = self.tedapi.get_native_meters_aggregates(force=force)
```
[the remark this answers is not in the document]
Comment (jasonacox-sam) [reply pypowerwall/tedapi/pypowerwall_tedapi.py:415]: Done — added `pypowerwall/tests/unit/test_native_energy_merge.py` (8 tests) covering both cases: (a) successful merge with the disclaimer provenance note appended per section, plus partial-payload and non-numeric-value edge cases, and (b) the unavailable path — retries suppressed for the 5-minute `NATIVE_FAIL_RETRY` window, retried after expiry, malformed payloads rejected, and successful fetches reused from cache on the normal poll cadence. Full suite: 285 passed, 10 skipped.

— Sam ⚡

**Comment by jasonacox-sam** [comment]
Done — all three items in e187d0f:

- **Review addressed:** added `pypowerwall/tests/unit/test_native_energy_merge.py` — 8 unit tests covering the successful merge + disclaimer provenance note, partial-payload and non-numeric edge cases, and the endpoint-unavailable backoff path (retry suppression within the 5-minute window, retry after expiry, malformed payload rejection, cache reuse). Full suite passes: 285 passed, 10 skipped.
- **Version:** bumped to `0.16.5`.
- **RELEASE.md:** v0.16.5 entry added.

Ready for merge whenever you are.

— Sam ⚡

**Comment by jasonacox** [comment]
@jasonacox-sam  how does this integrate with https://github.com/jasonacox/pypowerwall/pull/359?

**Comment by jasonacox-sam** [comment]
Short answer: they're orthogonal changes that compose cleanly — I merged both branches into a combined tree to verify, and the only thing that needs coordinating is the version number, not code.

**How the two PRs relate**

* **#359 (bearer auth mode)** changes how the *TEDAPI protobuf queries* authenticate: installer login (full gateway password from the QR sticker) → Bearer token → every query wrapped in an `AuthEnvelope` — enabling wired-LAN access without a static route on PW2 and solar-only systems (explicitly not PW3).
* **#372 (this PR)** adds a separate side channel: a *customer* login (password derived from the last 5 chars of the gateway password, running on its own `requests.Session`) that fetches the classic `/api/meters/aggregates` REST endpoint to overlay the lifetime energy accumulators. It runs regardless of how the TEDAPI session authenticates — basic, bearer, or v1r.

Both POST to `/api/login/Basic`, but with different credentials on different sessions, so there's no token or header interference. The energy merge behaves identically under bearer auth as under basic, and on PW3 wired (v1r) — the path #359 explicitly doesn't support — this PR is the one carrying the energy data. Across hardware the two are complementary.

**Verified against a combined tree**

* Merged `main` → this PR → #359 locally: all Python files (`pypowerwall/tedapi/__init__.py`, `pypowerwall_tedapi.py`, top-level `__init__.py`) auto-merge cleanly.
* The only conflict is `RELEASE.md` — both PRs add a `v0.16.5` section and bump `version_tuple`. Trivially resolved by sequencing.
* Combined test suite on the merged tree: **388 passed, 10 skipped** (vs 285 on this branch alone) — the bearer tests and the native-energy tests all pass together.

**Suggested sequencing**

Merge this PR as v0.16.5, then #359 rebases on top and takes v0.16.6 — a one-line `version_tuple` bump plus the RELEASE.md heading renumber, nothing else. (Reverse order works too; whichever merges second just renumbers.)

The earlier prep items are all in e187d0f — review tests, version bump, and RELEASE.md entry. Ready for merge whenever you are.

— Sam ⚡

**Comment by Nexarian** [comment]
I'm genuinely surprised we can get anything useful at all out of the /api/login/Basic login on a PW3 with the customer login.

@jasonacox-sam Questions:
- Does that produce a token as well? The way this is implemented, there's separate functions for customer and installer login.
- We might want to merge the login paths, if the only real functional difference is the customer versus the installer login, assuming they both produce bearer tokens, the only difference is the authorized scopes (what APIs work and what data you can get out of it). This is worth further deep-dives. I had abandoned the customer login path because I assumed it just didn't allow ANYTHING useful.
- I assume that this doesn't affect V2024_06 or V2026_06 TEDAPI

It also might be worth merging everything and then doing a refactoring pass. That's my favorite trick to move code forward, though I know @jasonacox hesistates on that ;)

**Comment by jasonacox-sam** [comment]
Good questions — each one verified against the live gateway (PW3, fw 26.18.1):

**1. Token? Yes.** `POST /api/login/Basic` with `username=customer` (password = last 5 chars of the gateway password) returns a JSON body with a `token`, used as a Bearer on subsequent GETs — same shape of flow as the installer login, different credential. Basic auth on the GET itself is a 403, so the token step is required, not optional. In this PR the token is cached and refreshed at 50 min (`CUSTOMER_TOKEN_EXPIRE = 3000s`; observed token lifetime ~1h), with a 300s backoff (`NATIVE_FAIL_RETRY`) if the endpoint refuses, so a gateway that won't serve it doesn't add a login POST to every poll.

**2. Merging the login paths.** Mechanically you're right — same endpoint, different username and credential source, and the difference is scope. What the customer token actually unlocks on current firmware is the open question: I've confirmed `/api/meters/aggregates` works (real lifetime Wh accumulators, live-ticking, identical over WiFi `192.168.91.1` and the v1r-wired gateway IP); anything beyond that endpoint is untested. Probing the rest of the classic `/api/*` surface with a customer token would be a genuinely useful deep-dive. One design note on why I *didn't* share the session with the TEDAPI side: the customer login has to coexist with whatever the main session is doing — basic, your bearer mode from #359, or v1r — without token or header interference, so it runs on its own `requests.Session` with its own cache and lock. Factoring a shared auth helper out is a reasonable post-merge cleanup once both branches land.

**3. V2024_06 / V2026_06 — correct, no effect.** The energy merge never touches the protobuf path: no `tedapi_pb2`, no `AuthEnvelope`, no interaction with `tedapi_api_version`. It's a plain HTTPS GET to the classic REST API on its own session, so it behaves identically whether queries use the legacy QueryType set or the V2026_06 GraphQL set — and on v1r wired it's the channel actually carrying the energy data, since that path doesn't ride the protobuf transport at all.

On the merge-then-refactor idea — the two PRs do compose cleanly (388 tests passing on a combined tree, only RELEASE.md needs renumbering for sequencing), so a consolidation pass after both land would start from a stable base.

— Sam ⚡

**Comment by jasonacox** [comment]
Took a closer pass at the transport plumbing behind `get_native_api()` / `get_native_meters_aggregates()` and found three issues worth fixing before merge, plus a couple of minors. The merge logic itself in `get_api_meters_aggregates()` checks out — fresh `API_METERS_AGGREGATES_STUB()` deepcopy per call so no disclaimer double-append, the cached native dict is only read and never mutated, and the degradation path is exactly prior behavior. The issues are all in the new fetch path:

## 1. `_customer_lock` is held across network calls, unbounded

`get_native_api()` wraps `_native_get()` in `with self._customer_lock:` — and `_native_get()` can make up to **four** network round trips (login POST + GET, ×2 on the auth retry), each bounded by `self.timeout`. Per AGENTS.md, locks shouldn't be held across network calls — and this one has *unbounded acquisition*: under the proxy's `ThreadingHTTPServer`, every concurrent `/api/meters/aggregates` request queues behind whoever is mid-fetch, with no timeout and no stale-cache escape hatch. Every other fetch in this class bounds lock acquisition (`acquire_lock_with_backoff`) and serves cached data on contention.

## 2. No double-check inside the lock → thundering herd on cache expiry

The cache/backoff check in `get_native_meters_aggregates()` runs *before* the lock is taken. When the ~5s `pwcacheexpire` window lapses with N concurrent clients, all N threads pass the stale-cache check and then serialize through the lock performing **N sequential full login+GET fetches** — the exact scenario the class's check-cache → acquire-lock → double-check pattern exists to prevent. This path skips the double-check.

## 3. Dead-LAN host ordering costs a full `timeout` on *every* poll

`get_native_api()` always tries `self.gw_ip` first, then `wifi_host`. In v1r mode with wired LAN down but WiFi up — a state the class already tracks via `self.lan_failed` — every fetch burns a full timeout on the dead host before succeeding over WiFi. And because the WiFi fetch *succeeds*, `_native_fail_until` never arms, so the penalty repeats on every cache expiry (~5s), not once per `NATIVE_FAIL_RETRY`. Notably `customer_host` already remembers the last host that worked (for token affinity) but isn't consulted for ordering.

## Minor

- `isinstance(value, (int, float))` in the merge accepts `bool` (`True` merges as `1`). Cheap to exclude.
- The module docstring's Functions list doesn't mention the new `get_native_api` / `get_native_meters_aggregates` methods (AGENTS.md convention).
- The codecov gap is real: tests mock at `get_native_api()`, leaving `_customer_login` / `_native_get` — including the 401 → re-login retry — untested. That path is cheap to cover with a mocked session.
- (Verified non-issues: `force=True` bypassing the backoff matches the established `force` semantics elsewhere in the class; the form-encoded login payload with nested `clientInfo` is byte-for-byte the proven `pypowerwall_local._get_session()` pattern.)

None of this changes the API surface — it's all behind the fetch boundary, so the new unit tests keep passing as-is. Fixes are small; pushing them to this branch shortly.

**Comment by jasonacox** [comment]
All four items fixed in 37dd9d0:

1. **Bounded lock acquisition** — `get_native_api()` / `get_native_meters_aggregates()` now acquire `_customer_lock` with `timeout=self.timeout`; on contention the meters path serves cached data and `get_native_api()` skips, so proxy threads never queue unbounded behind a multi-round-trip fetch.
2. **Double-checked cache** — cache/backoff re-checked after acquiring the lock (now an `RLock` so the meters path holds it across `get_native_api`); concurrent cache expiry triggers exactly one refetch.
3. **Host ordering** — the last host that served us (`customer_host`) is tried first, with `lan_failed` as the hint before any success; a dead wired LAN no longer costs a full timeout per poll when WiFi is carrying the data.
4. **Minors** — `bool` excluded from the numeric merge guard; new methods added to the module docstring Functions list; RELEASE.md notes the hardening.

Coverage gap closed too: 9 new tests exercise `_customer_login`/`_native_get` directly — 401 → re-login retry (including the fresh Bearer on the retried GET), tokenless and non-JSON responses, both host-ordering rules, lock-contention stale-serve, and the bool guard. `test_native_energy_merge.py` is now 17 tests; full suite 373 passed, 10 skipped.

No API surface changes — everything is behind the fetch boundary, and the original 8 tests pass unmodified.

**Comment by jasonacox-sam** [comment]
Verified `37dd9d0` against the live gateway (PW3, dual + expansion, fw 26.18.1, v1r wired) — the hardening holds up on real hardware.

**Full suite:** 373 passed, 10 skipped — matches.

**Live checks on the customer-login fetch path:**

| Check | Result |
|---|---|
| Energy merge (`/api/meters/aggregates`) | ✅ all four sections carry non-zero lifetime accumulators; disclaimer notes "energy from gateway local API" |
| Token reuse | ✅ forced uncached fetch performed **0** re-logins — cached Bearer honored |
| Concurrent burst (8 threads, cache expired) | ✅ exactly **1** network fetch; 8/8 threads served valid data in ~0.02s |
| Host affinity | ✅ `customer_host` pinned to the wired host after first success, so the dead-LAN reordering rule has its hint |

The burst test is the one unit tests can't quite reproduce with real latency — eight proxy threads hitting cache expiry simultaneously is exactly the `ThreadingHTTPServer` scenario from finding 1, and it went from N serialized login+GET round trips to one fetch plus cached serves for everyone else.

Ready for merge from my side.

— Sam ⚡

**Comment by cavedon** [comment]
That's great, thank you for the quick turn-around. Will these metrics be automatically pushed via MQTT?  If not, can that be added as well, please?
Update: I guess that would be a change for the powerwall-server project. Shall I open a ticket there?

**Comment by jasonacox-sam** [comment]
Short answer: one half flows automatically, the other half is exactly the pypowerwall-server change you're thinking of — so yes, please open the ticket there.

**What's automatic:** the server's MQTT publisher re-publishes the *full* `/api/meters/aggregates` payload verbatim on `{prefix}/{gateway_id}/aggregates` (alongside the scalar `solar` / `grid` / `home` / `powerwall` topics). That payload comes straight from the library, so once this PR ships and pypowerwall-server bumps its pinned dependency (currently `pypowerwall==0.16.2`), the lifetime `energy_imported` / `energy_exported` values will appear in that JSON topic with zero server-side code changes. If you consume the aggregates topic in Node-RED or similar, you're done at that point.

**What needs the ticket:** the curated scalar topics and the Home Assistant auto-discovery sensors only cover instantaneous power, battery SoE, mode/reserve, and solar strings — no lifetime energy sensors today. Adding per-section lifetime energy topics (site/battery/load/solar, imported + exported, Wh) plus the matching HA discovery entries is a small, well-contained change in `app/mqtt/publisher.py` and `app/mqtt/ha_discovery.py` — worth doing since HA users won't want to parse raw JSON just to graph cumulative energy. The ticket is also a good place to settle semantics: these are lifetime totals in Wh, same as PW2's counters, so daily-energy dashboards will want to delta them exactly as they already do for PW2.

And thanks again for surfacing this endpoint in #221 — none of this exists without that find.

For thread completeness: the review tests, hardening, version bump to 0.16.5, and RELEASE.md entry are all on the branch (e187d0f, 37dd9d0 — 373 tests passing, verified on live hardware), and the relationship to #359 is covered in the earlier reply above. Ready for merge.

— Sam ⚡

**Comment by jasonacox** [comment]
Thanks everyone!

View at:
https://pypi.org/project/pypowerwall/0.16.5/

@jasonacox-sam please create a new tag and github release for this version. Include changes since last tag and add any contributor mentions. Post on X when done.

**Comment by jasonacox-sam** [comment]
All done — release is live:

🎯 **Release:** https://github.com/jasonacox/pypowerwall/releases/tag/v0.16.5
📦 **PyPI:** https://pypi.org/project/pypowerwall/0.16.5/
🐦 **X:** https://x.com/jasonacox_sam/status/2089051462194397431

Tag `v0.16.5` sits on the merge (`e3c6acb`), and the release notes cover everything since `v0.16.4` — the lifetime energy accumulator merge, the thread-safety hardening, and the new test coverage — with a Contributors section crediting @cavedon for the endpoint discovery this whole feature is built on, @ajfhodgson for the original #221 report, and @Nexarian for the auth-path review.

— Sam ⚡

**Comment by jasonacox** [comment]
Legacy proxy: jasonacox/pypowerwall:0.16.5t99

**Comment by jasonacox-sam** [comment]
Confirmed on Docker Hub — `jasonacox/pypowerwall:0.16.5t99` and `latest` both live as of 17:58 UTC today, and I've added a **Proxy (t99)** section to the release notes so the image tag is discoverable there too.

That closes out every artifact in the v0.16.5 chain, each verified live:

* **Tag:** `v0.16.5` on merge `e3c6acb`
* **GitHub release:** https://github.com/jasonacox/pypowerwall/releases/tag/v0.16.5
* **PyPI:** https://pypi.org/project/pypowerwall/0.16.5/
* **X:** https://x.com/jasonacox_sam/status/2089051462194397431
* **Docker:** `0.16.5t99` + `latest` ✓

PW3 users who pull the new proxy image will see lifetime `energy_imported` / `energy_exported` in `/api/meters/aggregates` on the first poll after upgrade — cumulative-energy dashboards start filling in the same way they've always worked on PW2.

— Sam ⚡

**Comment by jasonacox** [comment]
@jasonacox-sam thanks, buddy! ❤️  Now, can you open a PR on https://github.com/jasonacox/pypowerwall-server to pull in this new library version and fully address the MQTT and HA support you mention above.

**Comment by jasonacox-sam** [comment]
Done — the pypowerwall-server PR is open: **https://github.com/jasonacox/pypowerwall-server/pull/75**

* **Library:** dependency bumped `0.16.2` → `0.16.5` (requirements.txt pin + pyproject floor; Docker images install from requirements.txt, so both pick it up automatically).
* **MQTT:** six lifetime energy topics (Wh) following the server's existing scalar convention (aggregates `site` → `grid`, `load` → `home`): `grid_energy_imported`, `grid_energy_exported`, `home_energy_imported`, `solar_energy_exported`, `battery_energy_imported`, `battery_energy_exported`. Topics mirror exactly what `/api/meters/aggregates` serves — gateways whose firmware lacks the native endpoint report `0`, nothing invented.
* **Home Assistant:** six matching auto-discovery sensors with `device_class: energy` + `state_class: total_increasing` (Wh), so they chart directly in the HA Energy dashboard — daily stats derived by delta, the same semantics PW2 has always had.

Full server test suite passes — 243 tests (7 new covering the energy topics and discovery payloads). Version bumped to `0.4.3` with a RELEASE.md entry in prep for merge.

And the question cavedon asked above is now fully covered end-to-end — the raw JSON on `{prefix}/{gw}/aggregates` via the library bump, the curated scalar topics + HA sensors via PR #75 — so no separate ticket needed.

— Sam ⚡
